display_list/display_list_color_source.h

```diff
@@ -0,0 +1,668 @@
+  static std::shared_ptr<DlColorSource> From(sk_sp<SkShader> sk_filter) {
+    return From(sk_filter.get());
+  }
+
+  static std::shared_ptr<DlColorSource> MakeLinear(
+      const SkPoint p0,
+      const SkPoint p1,
+      uint32_t stop_count,
+      const uint32_t* colors,
+      const float* stops,
+      DlTileMode tile_mode,
+      const SkMatrix* matrix = nullptr);
+
+  static std::shared_ptr<DlColorSource> MakeRadial(
+      SkPoint center,
+      SkScalar radius,
+      uint32_t stop_count,
+      const uint32_t* colors,
+      const float* stops,
+      DlTileMode tile_mode,
+      const SkMatrix* matrix = nullptr);
+
+  static std::shared_ptr<DlColorSource> MakeConical(
+      SkPoint start_center,
+      SkScalar start_radius,
+      SkPoint end_center,
+      SkScalar end_radius,
+      uint32_t stop_count,
+      const uint32_t* colors,
+      const float* stops,
+      DlTileMode tile_mode,
+      const SkMatrix* matrix = nullptr);
+
+  static std::shared_ptr<DlColorSource> MakeSweep(
+      SkPoint center,
+      SkScalar start,
+      SkScalar end,
+      uint32_t stop_count,
+      const uint32_t* colors,
+      const float* stops,
+      DlTileMode tile_mode,
+      const SkMatrix* matrix = nullptr);
+
+  virtual bool is_opaque() const = 0;
+
+  virtual std::shared_ptr<DlColorSource> with_sampling(
+      const SkSamplingOptions& options) const {
+    return shared();
+  }
+
+  // Return a DlColorColorSource pointer to this object iff it is an Color
+  // type of ColorSource, otherwise return nullptr.
+  virtual const DlColorColorSource* asColor() const { return nullptr; }
+
+  // Return a DlImageColorSource pointer to this object iff it is an Image
+  // type of ColorSource, otherwise return nullptr.
+  virtual const DlImageColorSource* asImage() const { return nullptr; }
+
+  // Return a DlLinearGradientColorSource pointer to this object iff it is a
+  // Linear Gradient type of ColorSource, otherwise return nullptr.
+  virtual const DlLinearGradientColorSource* asLinearGradient() const {
+    return nullptr;
+  }
+
+  // Return a DlRadialGradientColorSource pointer to this object iff it is a
+  // Radial Gradient type of ColorSource, otherwise return nullptr.
+  virtual const DlRadialGradientColorSource* asRadialGradient() const {
+    return nullptr;
+  }
+
+  // Return a DlConicalGradientColorSource pointer to this object iff it is a
+  // Conical Gradient type of ColorSource, otherwise return nullptr.
+  virtual const DlConicalGradientColorSource* asConicalGradient() const {
+    return nullptr;
+  }
+
+  // Return a DlSweepGradientColorSource pointer to this object iff it is a
+  // Sweep Gradient type of ColorSource, otherwise return nullptr.
+  virtual const DlSweepGradientColorSource* asSweepGradient() const {
+    return nullptr;
+  }
+
+ protected:
+  DlColorSource() = default;
+
+ private:
+  FML_DISALLOW_COPY_ASSIGN_AND_MOVE(DlColorSource);
+};
+
+class DlColorColorSource final : public DlColorSource {
```

Comment:
Perhaps `DlFlatColorSource`?

Reply:
Are these subclasses and others missing defaulted virtual destructors?

Reply:
"Flat" doesn't speak to me. The Skia equivalent is SkColorShader. I'll have to think about that.

There is a default virtual constructor in the base class (DlAttribute). I don't think I considered beyond that, so I'll check into the destructors.

Reply:
Sounds good.